library/std/src/sync/mpsc/mod.rs

```diff
@@ -1016,15 +745,14 @@ impl<T> SyncSender<T> {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T> Clone for SyncSender<T> {
     fn clone(&self) -> SyncSender<T> {
-        self.inner.clone_chan();
-        SyncSender::new(self.inner.clone())
+        SyncSender { inner: self.inner.clone() }
     }
 }
 
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T> Drop for SyncSender<T> {
     fn drop(&mut self) {
-        self.inner.drop_chan();
+        let _ = self.inner;
```

Comment:
Just curious, does this line do anything? In my understanding, we cannot move out of mutable reference, but Rust compiler will treat this line as no-op since it sees `let _`. The actual drop for inner is inserted by compiler somewhere else. This won't affect the code generation, but just a little confusing to me.

Reply:
`self.inner` becomes a statement temporary, and the drop glue is inserted at the end of the statement instead of the end of the function.

Reply:
You're right that impl doesn't actually do anything, but removing it would break `T: Drop` bounds, which I think would be a breaking change?

Reply:
`let _ = self.inner;` is basically obfuscated code for `drop(self.inner);` Wouldn't that be a more readable line of code to use here?

Reply:
However actually I think @riking you are wrong. `let _ = *ptr;` is a NOP. `let _ = expr;` is only equivalent to `drop(expr)` when `expr` is not a pure place expression.

If you want a destructor to be executed, just call `drop`.

Reply:
> So `let _ = self.inner;` is basically obfuscated code for `drop(self.inner);`? Wouldn't that be a more readable line of code to use here?

@RalfJung I think the problem is `drop(self.inner)` won't compile because `self.inner` cannot be moved out of a reference. The following code won't compile.

```rust
struct A { inner: String }
impl Drop for A {
    fn drop(&mut self) {
        drop(self.inner);
    }
}
fn main() {
    let _x = A { inner: "".to_string() };
}
```

In practice, `drop` for inner don't need to be explicitly called, and the compiler will insert it automatically. `let _ = self.inner;` can just be omitted, and leave the drop function an empty body.

Reply:
>  I think the problem is drop(self.inner) won't compile because self.inner cannot be moved out of a reference. The following code won't compile.


This demonstrates that `let _ = self.inner;` is indeed a NOP, and should be removed IMO.

Reply:
> This demonstrates that `let _ = self.inner;` is indeed a NOP, and should be removed IMO.

Agree

Reply:
Opened https://github.com/rust-lang/rust/pull/105243